### PR TITLE
Init specs for "sdk-wallet"

### PR DESCRIPTION
**Pull request by andreibancioiu**
For reviewers:
 - When reviewing `sdk-wallet/crypto`, make sure to have a look here, as well: https://github.com/multiversx/mx-sdk-js-wallet/tree/main/src/crypto

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+### **`pay-attention-to-types`**
+
+ - For JavaScript / TypeScript, `bytes` should be `Uint8Array`.
+
+### **`follow-language-conventions`**
+ 
+ - Make sure to follow the naming conventions of the language you're using, e.g. `snake_case` vs. `camelCase`.
+ - In the specs, interfaces are prefixed with `I`, simply to make them stand out. However, in the implementing libraries, this convention does not have to be applied.
+ - In `go`, the term `serialize` (whether it's part of a class name or a function name) can be replaced by `marshal`, since that is the convention.
+
```
Comment (popenta) [inline README.md:28]: Errors should also follow the language convention. Should have the same error message, though.
`ErrInvalidPublicKey` vs `InvalidPublicKeyError`.

File: sdk-core/address.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    // Named constructor
+    // Can throw:
+    // - ErrInvalidBech32Address
+    static new_from_bech32(value: string): Address;
+
+    // Named constructor
+    // Can throw:
+    // - ErrInvalidHexString
+    static new_from_hex(value: string, hrp: string): Address;
```
Comment (popenta) [inline sdk-core/address.md:18]: Maybe without `new_`?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Specifications for mx-sdk-* libraries
+
+This repository contains specifications for the `mx-sdk-*` libraries. The specifications are written in a language-agnostic manner, and are meant to be implemented in multiple languages (e.g. Go, TypeScript, Python, Rust, etc.).
```
Comment (iulianpascalau) [inline README.md:3]: :+1:

File: README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+- `sdk-wallet`: core wallet components (generation, signing).
+- `sdk-network-providers`: Network Provider (API, Gateway) components.
+
+## Guidelines
+
+###  **`in-ifaces-out-concrete-types`**
+
+Generally speaking, it's recommended to receive input parameters as abstractions (interfaces) in the public API of the SDKs. This leads to an improved decoupling, and allows for easier type substitution (e.g. easier mocking, testing).
+
+Generally speaking, it's recommended to return concrete types in the public API of the SDKs. The client code is responsible with decoupling from unnecessary data and behaviour of returned objects (e.g. by using interfaces, on their side).
```
Comment (iulianpascalau) [inline README.md:17]: We might add:
```
The only notable exception to this is when working with factories (abstract or method factories) that should make the function output an interface type
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+Generally speaking, it's recommended to return concrete types in the public API of the SDKs. The client code is responsible with decoupling from unnecessary data and behaviour of returned objects (e.g. by using interfaces, on their side). 
+
+### **`pay-attention-to-types`**
+
+ - For JavaScript / TypeScript, `bytes` should be `Uint8Array`.
+
+### **`follow-language-conventions`**
+ 
+ - Make sure to follow the naming conventions of the language you're using, e.g. `snake_case` vs. `camelCase`.
+ - In the specs, interfaces are prefixed with `I`, simply to make them stand out. However, in the implementing libraries, this convention does not have to be applied.
```
Comment (iulianpascalau) [inline README.md:26]: :+1:

File: sdk-core/address.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
+## Address
+
+```
+class Address:
+    // Should also validate the length of the provided input.
+    // Can throw:
```
Comment (iulianpascalau) [inline sdk-core/address.md:6]: :+1: :+1: :+1:

File: sdk-core/address.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+address = factory.create_from_bech32("erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th")
+address = factory.create_from_hex("0139472eff6886771a982f3083da5d421f24c29181e63888228dc81ca60d69e1")
+address = factory.create_from_public_key(bytes.fromhex("0139472eff6886771a982f3083da5d421f24c29181e63888228dc81ca60d69e1"))
+```
+
+Module-level functions:
+
+```
+// Note that the first input parameter is received as an interface, but the return value is a concrete type [see Guideline **`in-ifaces-out-concrete-types`**].
+compute_contract_address(deployer: IAddress, deployment_nonce: INonce): Address;
```
Comment (iulianpascalau) [inline sdk-core/address.md:82]: as stated in the other PR, we might think of a component that will contain all these helpers. Reason: to implement all these compute functions we need some inner components (shard coordinator, blockchain hook (!) and such). To overcome these continuous sub-components we might create them once (on the constructor) and use them whenever we want.

File: sdk-wallet/crypto/keypair_based_encryptor_decryptor.md [hunk 8]
```diff
@@ -0,0 +1,8 @@
+## KeyPairBasedEncryptorDecryptor
+
+```
+class KeyPairBasedEncryptorDecryptor:
+    encrypt(data: bytes, recipient_public_key: IPublicKey, auth_secret_key: ISecretKey): PublicKeyEncryptedData;
+
+    decrypt(data: PublicKeyEncryptedData, decryptor_secret_key: SecretKey): data;
```
Comment (iulianpascalau) [inline sdk-wallet/crypto/keypair_based_encryptor_decryptor.md:7]: decrypt returns data?

File: sdk-wallet/user_wallet_provider.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    // Can throw:
+    // - ErrInvalidSecretKey
+    compute_public_key_from_secret_key(secret_key: ISecretKey): IPublicKey
+
+    // Should not throw.
+    generate_mnemonic(): Mnemonic
+
+    // Can throw:
+    // - ErrInvalidMnemonic
+    derive_secret_key_from_mnemonic(mnemonic: Mnemonic, address_index: number = 0, passphrase: string = ""): ISecretKey
```
Comment (iulianpascalau) [inline sdk-wallet/user_wallet_provider.md:34]: I see this as part of the mnemonic definition. The possibility to convert itself to some other part. Do you think it is worth moving there and keeping only the `generate_mnemonic` function here?
we have L5 on mnemonic.md :D

File: sdk-wallet/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+## Keystores
+
+Components in `sdk-wallet/keystore`, even they do share a common purpose - storing secret keys - are not meant to be used interchangeably. Generally speaking, they do not (are not required to) share a common interface. If needed (for exotic use-cases), client code can design customized adapters over these components in order to unify their interfaces (this is not covered by specs).
```
Comment (bogdan-rosianu) [inline sdk-wallet/README.md:3]: even **if** they do share a common purpose*

File: README.md (hunk 1 above)
Comment (andreibancioiu) [reply README.md:28]: Added.

File: README.md (hunk 4 above)
Comment (andreibancioiu) [reply README.md:17]: Added.

File: sdk-wallet/crypto/keypair_based_encryptor_decryptor.md (hunk 8 above)
Comment (andreibancioiu) [reply sdk-wallet/crypto/keypair_based_encryptor_decryptor.md:7]: By mistake, fixed.

File: sdk-wallet/user_wallet_provider.md (hunk 9 above)
Comment (andreibancioiu) [reply sdk-wallet/user_wallet_provider.md:34]: Indeed, moved.

File: sdk-wallet/README.md (hunk 10 above)
Comment (andreibancioiu) [reply sdk-wallet/README.md:3]: Fixed.

File: sdk-core/address.md (hunk 7 above)
Comment (andreibancioiu) [reply sdk-core/address.md:82]: Not fixed yet, I am working on it.

**Review by ccorcoveanu — APPROVED** [review]

File: sdk-core/address.md (hunk 7 above)
Comment (andreibancioiu) [reply sdk-core/address.md:82]: Fixed, added `AddressComputer`.

**Review by bogdan-rosianu — APPROVED** [review]

File: sdk-wallet/user_wallet_provider.md [hunk 11]
```diff
@@ -5,6 +5,14 @@ class UserWalletProvider:
     // Should not throw.
     generate_keypair(): (ISecretKey, IPublicKey)
 
+    // Should not throw (generally speaking).
+    //
+    // Generates a keypair that satisfies the given constraints.
+    // 
+    // Generally speaking, the input paramters would be a predicate (function that returns a boolean) 
+    // or something similar (depending on the programming language).
+    generate_keypair_with_constraint(isConstraintSatisfied: Predicate[input = (ISecretKey, IPublicKey), output = boolean]): (ISecretKey, IPublicKey)
```
Comment (andreibancioiu) [inline sdk-wallet/user_wallet_provider.md:14]: Or perhaps this should be optional, not part of specs?

Comment (ccorcoveanu) [reply sdk-wallet/user_wallet_provider.md:14]: Yes, I think so, I would keep the specs simple and to the point

File: README.md [hunk 12]
```diff
@@ -14,7 +14,7 @@ This repository contains specifications for the `mx-sdk-*` libraries. The specif
 
 Generally speaking, it's recommended to receive input parameters as abstractions (interfaces) in the public API of the SDKs. This leads to an improved decoupling, and allows for easier type substitution (e.g. easier mocking, testing).
 
-Generally speaking, it's recommended to return concrete types in the public API of the SDKs. The client code is responsible with decoupling from unnecessary data and behaviour of returned objects (e.g. by using interfaces, on their side). 
+Generally speaking, it's recommended to return concrete types in the public API of the SDKs. The client code is responsible with decoupling from unnecessary data and behaviour of returned objects (e.g. by using interfaces, on their side). The only notable exception to this is when working with factories (abstract or method factories) that should have the function output an interface type. For example, have a look over `(User|Validator)WalletProvider.generate_keypair()` - this method returns abstract types (interfaces).
```
Comment (iulianpascalau) [inline README.md:17]: :+1:

**Review by iulianpascalau — DISMISSED** [review]

File: sdk-wallet/user_wallet_provider.md (hunk 11 above)
Comment (andreibancioiu) [reply sdk-wallet/user_wallet_provider.md]: All right, removed `generate_keypair_with_constraint` from specs.

File: sdk-core/address.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+print("Address (bech32-encoded)", address.bech32())
+print("Public key (hex-encoded):", address.hex())
+```
+
+## AddressFactory
+
+```
+class AddressFactory:
+    constructor(hrp: string = DEFAULT_HRP);
```
Comment (CiprianDraghici) [inline sdk-core/address.md:51]: What would be the `DEFAULT_HRP` value ?

**Review by CiprianDraghici — APPROVED** [review]

File: sdk-core/address.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+    // Named constructor
+    // Can throw:
+    // - ErrInvalidBech32Address
+    static new_from_bech32(value: string): Address;
+
+    // Named constructor
+    // Can throw:
+    // - ErrInvalidHexString
+    static new_from_hex(value: string, hrp: string): Address;
```
[the remark this answers is not in the document]
Comment (popenta) [reply sdk-core/address.md:18]: `hrp` stands for `human readable part` and in our case it's `erd` (the beginning of the address)

File: sdk-core/address.md (hunk 13 above)
Comment (andreibancioiu) [reply sdk-core/address.md:51]: Changed to use string literal.

File: sdk-core/address.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply sdk-core/address.md:18]: For the first comment - yes, in the TypeScript implementation we will use that approach.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by CiprianDraghici — APPROVED** [review]

**Review by popenta — APPROVED** [review]

**Review by ccorcoveanu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**